README.md
[![Build Status](https://travis-ci.org/mozilla/browserid-local-verify.png?branch=master)](https://travis-ci.org/mozilla/browserid-local-verify)

(status:  **experimental**)

# A node.js BrowserID verification library

This repository contains a node.js library for local verification of BrowserID assertions.

The library has the following scope and features:

  * **authoritative domain lookup** - given a domain, follow the browserid standard to resolve it (following authority delagation) into the ultimatly responsible domain and its public key.
  * **.well-known document parsing**
  * **(multiple) secondary IdP support** - The library can be initialized with a set of trusted "fallback IdPs" that are considered authoritative when lookup fails (no support document can be found).
  * **external HTTP implementation** - You can use node.js's http implementation, or override it and support your own (useful for HTTP proxied environments)
  * **command line tools** - all of these features are exposed via command line tools for manual inspection of domain's persona configuration.
  * **assertion verification** - the features above all fuel a simple yet flexible API for local verification of assertions.

# Alternatives

This library is targeted at robust local verification, to subsume all of the features required by mozilla's implementation of assertion verification in [Persona][].  If you're looking for a clean and simple library appropriate for website use (using the verifier hosted by persona), see [browserid-verify][].

[Persona]: https://persona.org
[browserid-verify]: https://npmjs.org/browserid-verify

# USAGE

    npm install persona-verifier-lib

## (simple) verifying an assertion

    var browserid = require('browserid-local-verify');

    browserid.verify({
      assertion: assertion,
      audience:  "http://example.com"
    }, function(err, details) {
      console.log(details);
    });

## looking up an authority for a domain

    var browserid = require('browserid-local-verify');

    browserid.lookup({
      domain: "mozilla.org"
    }, function(err, details) {
      // check err
      console.log(details.authority);
      console.log(details.pubKey);
      console.log(details.delegationChain);
    });

## configuration

All functions accept configuration parameters documented below.

    browserid.lookup({
      httpTimeout: 5.0,
      domain: "mozilla.org"
    }, function(err, details) {
      ...
    });

Or you can allocate a library instance.  This allows you to specify configuration once at instantiation time.  Any configuration parameters or function arguments may be specified a instantiation time and become the default for subsequently invoked functions:

    var BrowserID = require('browserid-local-verify');

    var b = new BrowserID({ httpTimeout: 20.0 });
    b.lookup({ domain: "mozilla.org" }, function(err, details) {
      // ...
    });

## Configuration and Arguments

### Common Arguments

* **httpRequest**: A function that allows the client to control how http requests are performed.
  * input arguments: (domain, path, callback)
  * callback argbuments: (err, statusCode, headers, body)
* **httpTimeout**: How long in seconds we should wait for a response when looking up a well-known document over HTTP. (default: 10)
* **maxDelegations**: How many times authority may be delegated.
* **insecureSSL**: When true, invalid SSL certificates are ignored (NEVER use this in production).
* **fallback**: A domain that is authoritative when support document lookup fails for the prinicpal email address's domain.

### lookup specific

* **domain**: the domain for which we should lookup the support document
* **principalDomain**: the domain of the email address for which we should discover the support document of the authority

### verification specific

* **now**: override the current time for purposes of assertion verification. (useful for testing)
* **assertion**: the assertion to verify
* **audience**: the expected assertion audience
* **trustedIssuers**: An array of domains that will be trusted to vouch for any identity, regardless of the authority as determined from the email addresses domain.

## debug output and metrics

The BrowserID class emits events:

    var b = new BrowserID();

    b.on('debug', function(msg) {
      console.log('debug output:', msg);
    });

    b.on('metric', function(metric, value) {
      console.log(metric + ":", value);
    });

    b.lookup("mozilla.org", function(err, details) {
      // ...
    });


ChangeLog
v0.0.9 -
	* support public-key change to pubkey
	* always specify algorithm as RSA or DSA

v0.0.8 -
	* augment 'tests/lib/client', allow caller to control whether .sub is present in cert

v0.0.7 -
	* support for "new data formats"

v0.0.6 -
	* bugfix: .fallback as an argument to .verify makes it through to .lookup for sloppy support docs.

v0.0.5 -
	* client testing tool now allows over-riding of principal

v0.0.4 -
	* now by default only the fallback is exempt from providing auth/prov urls
	* improved code instrumentation
	* improved error messages

v0.0.3 -
	* require /testing to include tools to facilitate automated unit testing

v0.0.2 -
	* fix sporadic travis failures
	* in code rename jwtRegisteredClaimNames to registeredClaimNames
	* library outputs userClaims instead of uaClaims

v0.0.1 -
	* Initial version.


tests/jshint.js
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

/* global describe,it */

// syntax checking of the source

const
should = require('should'),
fs = require('fs'),
path = require('path'),
jshint = require('jshint').JSHINT,
walk = require('walk'),
async = require('async'),
util = require('util');

describe('source code syntax', function() {
  // read jshintrc
  var jshintrc;

  it('.jshintrc should be readable', function(done) {
    jshintrc = JSON.parse(fs.readFileSync(path.join(__dirname, '../.jshintrc')).toString());
    (jshintrc).should.be.type('object');
    done();
  });

  var filesToLint = [
    path.join(__dirname, '../index.js')
  ];

  it('we should be able to discover files to lint', function(done) {
    async.each([
      path.join(__dirname, '../lib'),
      __dirname
    ], function(dir, done) {
      var walker = walk.walk(dir, {});

      walker.on("file", function(root, fStat, next) {
        var f = path.join(root, fStat.name);
        if (/\.js$/.test(f)) {
          filesToLint.push(f);
        }
        next();
      });
      walker.on("end", done);
    }, done);
  });

  it('syntax checking should yield no errors', function(done) {
    var errors = [];

    function checkNext() {
      if (!filesToLint.length) {
        if (errors.length) {
          var buf = util.format("\n        %d errors:\n        * ",
                                errors.length);
          buf += errors.join("\n        * ");
          done(buf);
        } else {
          done(null);
        }
        return;
      }
      var f = filesToLint.shift();
      fs.readFile(f.toString(), function(err, data) {
        should.not.exist(err);
        f = path.relative(process.cwd(), f);
        if (!jshint(data.toString(), jshintrc)) {
          jshint.errors.forEach(function(e) {
            errors.push(util.format("%s %s:%d - %s", e.id, f, e.line, e.reason));
          });
        }
        checkNext();
      });
    }
    checkNext();
  });
});


tests/lookup.js
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

/* global describe,it,require */

const
url = require('url'),
should = require('should'),
BrowserID = require('../'),
IdP = require('./lib/idp.js'),
Proxy = require('./lib/proxy.js'),
async = require('async');

describe('.well-known lookup transport tests (HTTP)', function() {
  // a local idp with a 1s delay in serving support documents
  var slowidp = new IdP({
    delay: 1.0
  });

  // a local idp that serves disabled support documents
  var disabledidp = new IdP({
    disabled: true,
  });

  // an idp that HTTP redirects
  var redirectidp = new IdP({
    http_redirect: 'example.com'
  });


  // a client library instance which ignores invalid SSL certs, and
  // only tolerates a 100ms delay in HTTP requests
  var browserid = new BrowserID({
    insecureSSL: true,
    httpTimeout: 0.1
  });

  // A client library instance that over-rides the built in HTTP
  // implementation
  var overRiddenBrowserid = new BrowserID({
    httpRequest: function(domain, path, cb) {
      cb(null, 200, { 'Content-Type': 'application/json' } , '{ "disabled": true }');
    }
  });

  var badContentType = new BrowserID({
    httpRequest: function(domain, path, cb) {
      cb(null, 200, { 'Content-Type': 'text/plain' } , '{ "disabled": true }');
    }
  });

  // A local proxy server.
  var proxy = new Proxy();

  it('test servers should start up', function(done) {
    async.parallel([
      function(cb) {
        slowidp.start(cb);
      },
      function(cb) {
        disabledidp.start(cb);
      },
      function(cb) {
        redirectidp.start(cb);
      },
      function(cb) {
        proxy.start(cb);
      }
    ], done);
  });

  it('should work with the built-in HTTP implementation', function(done) {
    browserid.lookup({ domain: disabledidp.domain() }, function(err, details) {
      should.not.exist(err);
      details.disabled.should.equal(true);
      details.delegationChain.should.be.instanceof(Array).and.have.lengthOf(1);
      details.delegationChain[0].should.equal(disabledidp.domain());
      details.authoritativeDomain.should.equal(disabledidp.domain());
      done(err);
    });
  });

  it('should work an over-ridden HTTP implementation', function(done) {
    overRiddenBrowserid.lookup({ domain: 'example.com' }, function(err, details) {
      should.not.exist(err);
      details.disabled.should.equal(true);
      details.delegationChain.should.be.instanceof(Array).and.have.lengthOf(1);
      details.delegationChain[0].should.equal('example.com');
      details.authoritativeDomain.should.equal('example.com');
      done(err);
    });
  });

  it('should timeout for slow http responses', function(done) {
    browserid.lookup({ domain: slowidp.domain() }, function(err) {
      should.exist(err);
      err.should.startWith('timeout trying to load well-known for 127.0.0.1:');
      done(null);
    });
  });

  it('should refuse to follow http redirects', function(done) {
    browserid.lookup({ domain: redirectidp.domain() }, function(err) {
      should.exist(err);
      err.should.endWith('is not a browserid primary - redirection not supported for support documents');
      done(null);
    });
  });

  it('should fail on wrong content type', function(done) {
    badContentType.lookup({ domain: redirectidp.domain() }, function(err) {
      should.exist(err);
      (err).should.contain('non "application/json" response');
      done(null);
    });
  });

  it('should use proxy server when $https_proxy is set', function(done) {
    var origHttpsProxy = process.env.https_proxy || '';
    process.env.https_proxy = proxy.url();
    proxy.clearNumRequests();
    browserid.lookup({ domain: disabledidp.domain() }, function(err, details) {
      should.not.exist(err);
      details.disabled.should.equal(true);
      should(proxy.numRequests()).equal(1);
      process.env.https_proxy = origHttpsProxy;
      done(err);
    });
  });

  it('should use proxy server when $HTTPS_PROXY is set', function(done) {
    var origHttpsProxy = process.env.HTTPS_PROXY || '';
    process.env.HTTPS_PROXY = proxy.url();
    proxy.clearNumRequests();
    browserid.lookup({ domain: disabledidp.domain() }, function(err, details) {
      should.not.exist(err);
      details.disabled.should.equal(true);
      should(proxy.numRequests()).equal(1);
      process.env.HTTPS_PROXY = origHttpsProxy;
      done(err);
    });
  });

  it('should not use proxy server when $http_proxy is set', function(done) {
    var origHttpProxy = process.env.http_proxy || '';
    process.env.http_proxy = proxy.url();
    proxy.clearNumRequests();
    browserid.lookup({ domain: disabledidp.domain() }, function(err, details) {
      should.not.exist(err);
      details.disabled.should.equal(true);
      should(proxy.numRequests()).equal(0);
      process.env.http_proxy = origHttpProxy;
      done(err);
    });
  });

  it('should bypass proxy server when $no_proxy is set to *', function(done) {
    var origHttpsProxy = process.env.https_proxy || '';
    var origNoProxy = process.env.no_proxy || '';
    process.env.https_proxy = proxy.url();
    process.env.no_proxy = '*';
    proxy.clearNumRequests();
    browserid.lookup({ domain: disabledidp.domain() }, function(err, details) {
      should.not.exist(err);
      details.disabled.should.equal(true);
      should(proxy.numRequests()).equal(0);
      process.env.no_proxy = origNoProxy;
      process.env.https_proxy = origHttpsProxy;
      done(err);
    });
  });

  it('should bypass proxy server when host is in $no_proxy', function(done) {
    var origHttpsProxy = process.env.https_proxy || '';
    var origNoProxy = process.env.no_proxy || '';
    var idpHostname = url.parse(disabledidp.url()).hostname;
    process.env.https_proxy = proxy.url();
    process.env.no_proxy = 'example.com, ' + idpHostname;
    proxy.clearNumRequests();
    browserid.lookup({ domain: disabledidp.domain() }, function(err, details) {
      should.not.exist(err);
      details.disabled.should.equal(true);
      should(proxy.numRequests()).equal(0);
      process.env.no_proxy = origNoProxy;
      process.env.https_proxy = origHttpsProxy;
      done(err);
    });
  });

  it('should use proxy server when host is not in $no_proxy', function(done) {
    var origHttpsProxy = process.env.https_proxy || '';
    var origNoProxy = process.env.no_proxy || '';
    process.env.https_proxy = proxy.url();
    process.env.no_proxy = 'example1.com, example1.com';
    proxy.clearNumRequests();
    browserid.lookup({ domain: disabledidp.domain() }, function(err, details) {
      should.not.exist(err);
      details.disabled.should.equal(true);
      should(proxy.numRequests()).equal(1);
      process.env.no_proxy = origNoProxy;
      process.env.https_proxy = origHttpsProxy;
      done(err);
    });
  });

  it('should fail on invalid domains', function(done) {
    browserid.lookup({ domain: 'http://example.com' }, function(err) {
      should.exist(err);
      (err).should.startWith('invalid domain:');
      done(null);
    });
  });

  it('test servers should shut down', function(done) {
    async.parallel([
      function(cb) {
        slowidp.stop(cb);
      },
      function(cb) {
        disabledidp.stop(cb);
      },
      function(cb) {
        redirectidp.stop(cb);
      },
      function(cb) {
        proxy.stop(cb);
      }
    ], done);
  });
});


tests/lib/idp.js
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

/* This file contains a test identity provider (IdP).  An IdP consists of,
 * for the purposes of testing verification, an SSL server (with self signed
 * certificate), and a keypair.  The servers are bound on ephemeral ports and
 * will be programatically configurable.  This allows us to robustly test
 * delegation chains, signing, and a whole bunch of other features implemented
 * by the verification library.
 */

/* jshint curly: false */

const
async = require('async'),
https = require('https'),
fs = require('fs'),
jwcrypto = require('jwcrypto'),
path = require('path');

// I hate this.
require("jwcrypto/lib/algs/rs");
require("jwcrypto/lib/algs/ds");

function IdP(args) {
  if (!args) args = {};
  this.args = args;
  // default paramter values
  this.args.delay = this.args.delay || 0;
  this.args.algorithm = this.args.algorithm || "rsa";
  this.args.keysize = this.args.keysize || 128;
  this.args.delegation = this.args.delegation || null;

  // allow algorithm specification as (i.e.) 'rsa' or 'RS'
  this.args.algorithm = this.args.algorithm.toUpperCase();

  if (args.wellKnown) this.wellKnown(args.wellKnown);
}

function later(cb /* args */) {
  var args = Array.prototype.slice.call(arguments, 1);
  process.nextTick(function() {
    cb.apply(null, args);
  });
}

IdP.prototype.url = function() {
  if (!this._started) throw "IdP isn't started, it has no url";
  return this.details.url;
};

IdP.prototype.publicKey = function() {
  if (!this._started) throw "IdP isn't started, it has no public key";
  return this.details.publicKey;
};

IdP.prototype.privateKey = function() {
  if (!this._started) throw "IdP isn't started, it has no public key";
  return this._secretKey;
};

IdP.prototype.domain = function() {
  if (!this._started) throw "IdP isn't started, it has no domain";
  return this.details.domain;
};

// the domain to whom this domain should delegate
IdP.prototype.delegation = function(domain) {
  if (domain === null) this.args.delegation = null;
  return (this.args.delegation = domain || this.args.delegation);
};

// the domain to whom this domain should delegate
IdP.prototype.delay = function(delay) {
  if (typeof delay === 'number') this.args.delay = delay;
  return this.args.delay;
};

// the domain to whom this domain should delegate
IdP.prototype.wellKnown = function(obj) {
  if (obj !== undefined) this.args.wellKnown = obj;
  return this.args.wellKnown || {
    authentication: '/auth.html',
    provisioning: '/prov.html',
    "public-key": this.publicKey().toSimpleObject()
  };
};


IdP.prototype.start = function(cb) {
  if (this._started) return later(cb, null, this.details);
  this._started = true;

  var self = this;

  function handleRequest(req, res) {
    if (req.url.indexOf('/.well-known/browserid') !== 0) {
      return res.send(404);
    }

    // XXX: these config values should all be mutually exclusive.  There's this odd
    // precedence going on that will confuse some poor test writer.
    if (self.args.http_redirect) {
      var location = 'https://' + self.args.http_redirect + '/.well-known/browserid';
      res.writeHead(301, {'Location': location});
      res.end();
    } else if (self.args.disabled) {
      res.writeHead(200, {'Content-Type': 'application/json'});
      res.end(JSON.stringify({disabled: true}));
    } else if (self.args.delegation) {
      res.writeHead(200, {'Content-Type': 'application/json'});
      res.end(JSON.stringify({authority: self.args.delegation}));
    }
    // if the caller supplied a 'dynamicWellKnown' function, allow them to
    // craft a support document after considering the URL with which it was
    // fetched
    else if (self.args.dynamicWellKnown) {
      self.args.dynamicWellKnown(req.url, function(err, doc) {
        if (err) {
          res.writeHead(500, {'Error': err});
        } else {
          res.writeHead(200, {'Content-Type': 'application/json'});
          res.end(JSON.stringify(doc || self.wellKnown()));
        }
      });
    } else {
      res.writeHead(200, {'Content-Type': 'application/json'});
      res.end(JSON.stringify(self.wellKnown()));
    }
  }

  async.parallel([
    function(cb) {
      // spin up an HTTPS server bound to an ephemeral port
      // using self signed certificates
      self._server = https.createServer({
        key: fs.readFileSync(path.join(__dirname, '..', 'resources', 'key.pem')),
        cert: fs.readFileSync(path.join(__dirname, '..', 'resources', 'cert.pem'))
      }, function (req, res) {
        setTimeout(function() {
          handleRequest(req, res);
        }, self.args.delay * 1000);
      }).listen(0, '127.0.0.1', function() {
        cb(null);
      });
    },
    function(cb) {
      // generate an RSA keypair for the idp
      jwcrypto.generateKeypair({
        algorithm: self.args.algorithm,
        keysize: self.args.keysize
      }, function(err, kp) {
        if (err) return cb(err);
        self._publicKey = kp.publicKey;
        self._secretKey = kp.secretKey;
        cb(null);
      });
    }
  ], function(err) {
    var addy = self._server.address();
    var domain = addy.address + ":" + addy.port;
    self.details = {
      url: "https://" + domain + "/",
      domain: domain,
      publicKey: self._publicKey
    };
    cb(err, self.details);
  });
};

IdP.prototype.stop = function(cb) {
  if (!this._started) return later(cb, null);
  else this._server.close(cb);
};

module.exports = IdP;


lib/compare-audiences.js
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

const
url = require('url');

// compare two audiences:
//   *want* is what was extracted from the assertion (it's trusted, we
//   generated it!
//   *got* is what was provided by the RP, so depending on their implementation
//   it might be strangely formed.
function compareAudiences(want, got) {
  function normalizeParsedURL(u) {
    // return object because url.parse returns object
    if (!u.port) {
      u.port = u.protocol === 'https:' ? '443' : '80';
    }
    return u;
  }

  try {
    var got_protocol, got_hostname, got_port;

    // We allow the RP to provide audience in multiple forms (see issue #82).
    // The RP SHOULD provide full origin, but we allow these alternate forms for
    // some dude named Postel doesn't go postal.
    // 1. full origin 'http://rp.tld'
    // 1a. full origin with port 'http://rp.tld:8080'
    // 2. domain and port 'rp.tld:8080'
    // 3. domain only 'rp.tld'

    // (app:// urls are seen on FirefoxOS desktop and possibly mobile)
    var gu = normalizeParsedURL(url.parse(got));
    if (gu.protocol !== 'https:' && gu.protocol !== 'http:' && gu.protocol !== 'app:') {
      // cases 2 & 3 default to http
      gu = normalizeParsedURL(url.parse('http://' + got));
    }
    got_protocol = gu.protocol;
    got_hostname = gu.hostname;
    got_port = gu.port;

    // now parse "want" url
    want = normalizeParsedURL(url.parse(want));

    if (got_protocol !== want.protocol) {
      throw "scheme mismatch";
    }
    if (got_port !== want.port) {
      throw "port mismatch";
    }
    if (got_hostname !== want.hostname) {
      throw "domain mismatch";
    }

    return undefined;
  } catch(e) {
    return e.toString();
  }
}

module.exports = compareAudiences;


lib/well-known-parser.js
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

var jwcrypto = require("jwcrypto");

// parse a well-known document.  throw an exception if it is invalid, return
// a parsed version if it is valid.  return is an object having the following fields:
//   * 'type' - one of "disabled", "delegation", or "supported"
//   * if type is "delegation", also:
//     * authority - the domain authority is delegated to
//   * if type is "supported":
//     * publicKey - a parsed representation of the public key
//     * paths.authentication - the path to the 'authentication' html
//     * paths.provisioning - the path to the 'provisioning' html
module.exports = function(doc, allowURLOmission) {
  try {
    doc = JSON.parse(doc);
  } catch(e) {
    throw "declaration of support is malformed (invalid json)";
  }

  if (typeof doc !== 'object') {
    throw "support document must contain a json object";
  }

  // there are three main types of support documents
  // 1. "supported" - declares the domain is a browserid authority,
  //    contains public-key, authentication, and provisioning
  // 2. "delegation" - declares the domain allows a different domain
  //    to be authoritative for it.
  // 3. "disable" - domain declares explicitly that it wants a secondary
  //    to be authoritative.

  // is this a "disable" document?
  if (doc.disabled === true) {
    return { type: "disabled" };
  } else if (doc.disabled !== undefined && doc.disabled !== false) {
    throw new Error("disabled must be either true or false");
  }

  // is this a delegation document?
  if (doc.authority) {
    if (typeof doc.authority !== 'string') {
      throw "malformed authority";
    }

    return {
      type: "delegation",
      authority: doc.authority
    };
  }

  // is this a support document?

  // the response that we'll populate as we go
  var parsed = {
    type: "supported",
    paths: {},
    publicKey: null
  };

  [ 'authentication', 'provisioning' ].forEach(function(requiredKey) {
    if (typeof doc[requiredKey] !== 'string') {
      if (!allowURLOmission) {
        throw "support document missing required property: '" + requiredKey + "'";
      }
    } else {
      parsed.paths[requiredKey] = doc[requiredKey];
    }
  });

  if (!doc['public-key']) {
    throw "support document missing required 'public-key'";
  }

  // can we parse that key?
  try {
    parsed.publicKey = jwcrypto.loadPublicKeyFromObject(doc['public-key']);
  } catch(e) {
    throw "mal-formed public key in support doc: " + e.toString();
  }

  // success!
  return parsed;
};


lib/browserid-local-verify.js
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

var
util = require("util"),
events = require("events"),
_ = require('underscore');

function Verifier(args) {
  events.EventEmitter.call(this);
  this.args = _.extend({
    'maxDelegations': 5,
    'httpTimeout': 10.0,
    'insecureSSL': false
  }, args);
}

util.inherits(Verifier, events.EventEmitter);

var lookup = require('./lookup.js');
Verifier.prototype.lookup = function(args, cb) {
  if (!cb) {
    cb = args;
    args = {};
  }
  lookup(this, _.extend({}, this.args, args), cb);
};

var verify = require('./verify.js');
Verifier.prototype.verify = function(args, cb) {
  if (!cb) {
    cb = args;
    args = {};
  }
  verify(this, _.extend({}, this.args, args), cb);
};

module.exports = Verifier;

module.exports.lookup = function(args, cb) {
  if (!cb) {
    cb = args;
    args = {};
  }
  var v = new Verifier(args);
  v.lookup(cb);
};

module.exports.verify = function(args, cb) {
  var v = new Verifier(args);
  v.verify(cb);
};

// expose utility function for testing
module.exports.compareAudiences = require('./compare-audiences');


lib/lookup.js
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

var
http = require('http'),
https = require('https'),
urlparse = require('urlparse'),
validation = require('./validation.js'),
wellKnownParser = require('./well-known-parser.js');

const WELL_KNOWN_URL = "/.well-known/browserid";

const DEFAULT_PORTS = {
  'http': 80,
  'https': 443
};

// like https.get() but transparently supports the
// $https_proxy and $no_proxy environment variables.
var getWithTransparentProxying = function(options, cb) {
  var httpmod = https;
  var proxy = shouldUseProxy(options.host);
  if (proxy) {
    if (proxy.scheme === 'http') {
      httpmod = http;
    }
    if (!options.headers) {
      options.headers = {};
    }
    if (!options.headers.host) {
      options.headers.host = options.host;
    }
    options.path = 'https://' + options.host + ':' + options.port + options.path;
    options.host = proxy.host;
    options.port = proxy.port || DEFAULT_PORTS[proxy.scheme];
  }
  return httpmod.get(options, cb);
};

var shouldUseProxy = function(host) {
  // Check if we have a proxy configured for https urls.
  var httpsProxy = process.env.https_proxy || process.env.HTTPS_PROXY;
  if (!httpsProxy) {
    return null;
  }
  httpsProxy = urlparse(httpsProxy).validate();
  // Check if this host is exclued via the no-proxy list.
  var noProxy = process.env.no_proxy || process.env.NO_PROXY;
  if (noProxy) {
    if (noProxy === '*') {
      return null;
    }
    var noProxyList = noProxy.split(/,\s*/);
    for (var i = 0; i < noProxyList.length; i++) {
      var suffix = noProxyList[i];
      if (host.lastIndexOf(suffix) === host.length - suffix.length) {
        return null;
      }
    }
  }
  // Yep, we need to use the proxy.
  return httpsProxy;
};


// hit the network and fetch a .well-known document in its unparsed form
var fetchWellKnown = function (emitter, args, currentDomain, principalDomain, clientCB) {
  // in many cases the http layer can send both an 'error' and an 'end'.  In
  // other cases, only 'error' will be emitted.  We want to
  // ensure the client callback is invoked only once.  this function does it.
  var startTime = new Date();
  var cb = function(err) {
    var reqTime = new Date() - startTime;

    emitter.emit('metric',
                 'elapsed_time.fetch_well_known.' + (err ? 'error' : 'success'),
                 reqTime);

    if (clientCB) {
      clientCB.apply(null, arguments);
      clientCB = null;
    }
  };

  function handleResponse(err, statusCode, headers, body) {
    if (statusCode !== 200) {
      if ([301, 302, 303, 307].indexOf(statusCode) !== -1) {
        return cb(currentDomain +
                  ' is not a browserid primary - redirection not supported for support documents');
      } else {
        return cb(currentDomain +
                  ' is not a browserid primary - non-200 response code to ' +
                  WELL_KNOWN_URL);
      }
    }
    // favor Postel over node convention
    var contentType = headers['content-type'] || headers['Content-Type'] || headers['Content-type'];
    if (!contentType || contentType.indexOf('application/json') !== 0) {
      return cb(currentDomain +
                ' is not a browserid primary - non "application/json" response to ' +
                WELL_KNOWN_URL);
    }

    cb(null, body, currentDomain);
  }

  var pathToWellKnown = WELL_KNOWN_URL + "?domain=" + principalDomain;

  emitter.emit('debug', 'fetching well-known from ' +  currentDomain + ' for principal ' +
               principalDomain);

  // If the client has provided a function to perform their own http requests,
  // use that.
  if (args.httpRequest) {
    return args.httpRequest(currentDomain, pathToWellKnown, handleResponse);
  } else {
    var port = DEFAULT_PORTS.https;
    var host = currentDomain;
    // somewhat odd, but we do allow a host:port to be specified as a domain.
    // this supports local testing using the built in https implementation.
    if (currentDomain.indexOf(':') !== -1) {
      var s = currentDomain.split(':');
      host = s[0];
      port = s[1];
    }
    var req = getWithTransparentProxying({
      host: host,
      port: port,
      path: pathToWellKnown,
      rejectUnauthorized: !args.insecureSSL,
      agent: false
    }, function(res) {
      var body = "";
      res.on('data', function(chunk) { body += chunk; });
      res.on('end', function() {
        handleResponse(null, res.statusCode, res.headers, body);
      });
    });
    req.end();

    // front-end shows xhr delay message after 10 sec; timeout sooner to avoid this
    var reqTimeout = setTimeout(function() {
      req.abort();
      return cb('timeout trying to load well-known for ' + currentDomain);
    }, args.httpTimeout * 1000);
    req.on('response', function() {
      if (reqTimeout) {
        clearTimeout(reqTimeout);
        reqTimeout = null;
      }
    });
    req.on('error', function(e) {
      if (reqTimeout) {
        clearTimeout(reqTimeout);
        reqTimeout = null;
      }
      return cb(currentDomain + ' is not a browserid primary: ' + String(e));
    });
  }
};

// Fetch a .well-known file from the network, following delegation
function lookup(emitter, args, currentDomain, principalDomain, cb, delegationChain) {
  if (!currentDomain) {
    currentDomain = principalDomain;
  }
  if (!principalDomain) {
    principalDomain = currentDomain;
  }
  if (!delegationChain) {
    delegationChain = [ principalDomain ];
  }

  try {
    validation.validateAuthority(principalDomain);
  } catch (e) {
    return cb("invalid domain: " + principalDomain);
  }
  if (currentDomain !== principalDomain) {
    try {
      validation.validateAuthority(currentDomain);
    } catch (e) {
      return cb("invalid domain: " + currentDomain);
    }
  }

  fetchWellKnown(emitter, args, currentDomain, principalDomain, function(err, unparsedDoc) {
    if (err) {
      return cb(err);
    }

    var supportDoc;
    try {
      var allowURLOmission = args.allowURLOmission || currentDomain === args.fallback;
      supportDoc = wellKnownParser(unparsedDoc, allowURLOmission);
    } catch (e) {
      return cb("bad support document for '" + currentDomain + "': " + String(e));
    }

    if (supportDoc.type === 'disabled')
    {
      return cb(null, {
        disabled: true,
        delegationChain: delegationChain,
        authoritativeDomain: delegationChain[delegationChain.length - 1],
      });
    }
    else if (supportDoc.type === 'delegation')
    {
      currentDomain = supportDoc.authority;

      // check for cycles in delegation
      if (delegationChain.indexOf(currentDomain) !== -1) {
        return cb("Circular reference in delegating authority: " + delegationChain.join(" > "));
      }

      delegationChain.push(currentDomain);

      emitter.emit('debug', delegationChain[delegationChain.length - 2] + " delegates to " +
                   delegationChain[delegationChain.length - 1]);

      // check for max delegation length (max delegations of n, implies we can have a chain
      // length of n + 1.  IOW, a chain of length 10, has 9 authority delegations.
      if (delegationChain.length > (args.maxDelegations + 1)) {
        return cb("Too many hops while delegating authority: " + delegationChain.join(" > "));
      }

      // recurse
      return lookup(emitter, args, currentDomain, principalDomain, cb, delegationChain);
    }
    else if (supportDoc.type === 'supported')
    {
      var url_prefix = 'https://' + currentDomain;

      var details = {
        publicKey: supportDoc.publicKey,
        delegationChain: delegationChain,
        authoritativeDomain: delegationChain[delegationChain.length - 1],
        urls: {
        }
      };

      // the well know parser has verified that urls are present unless
      // allowURLOmission is specified.  Now we'll validate urls are
      // correct if they have been supplied
      try {
        if (supportDoc.paths.authentication) {
          details.urls.auth = url_prefix + supportDoc.paths.authentication;
        }
        if (supportDoc.paths.provisioning) {
          details.urls.prov = url_prefix + supportDoc.paths.provisioning;
        }
      } catch(e) {
        return cb("invalid URL in support document: " + e.toString());
      }

      // success!
      cb(null, details);
    }
    else
    {
      var msg = "unhandled error while parsing support document for " + currentDomain;
      return cb(msg);
    }
  });
}

module.exports = function (browserid, args, cb) {
  lookup(browserid, args, args.domain, args.principalDomain, function(err, details) {
    // if there is an error, then let's try the fallback if configured
    if (err && args.fallback) {
      return lookup(browserid, args, args.fallback, args.principalDomain, cb, [ args.fallback ]);
    }
    cb(err, details);
  });
};
